### PR TITLE
Start the split of the Gallina Extensions chapter.

### PR DESCRIPTION
**Pull request by Zimmi48**
**Kind:** documentation

As part of the ongoing refactoring of the reference manual, and now that the grammar in this chapter has been updated, I started to split out this very long chapter into several sections that are moved to the place where they are most relevant.

Like this was the case for #11871, there will be a separate PR doing the same split for v8.11 but without any visible change.

The first proposal in this PR is to move the section on records out into the new "Core language" chapter. Right now, it is a bit alone but it is soon going to be joined by a section on "Variants and the match construct" which will come before "Records" and two sections on "Inductive types and recursive functions" and "Co-inductive types and co-recursive functions". Right now records are described as "macros" but the goal will be to rewrite this to present this as something more "primitive" (though this rewording will come separately from this splitting PR).

The second proposal in this PR is to move the section on "Sections" out into the "Core language" chapter, just before the "Modules" chapter. The "Modules" chapter will eventually be consolidated with the section on modules from the Gallina Extensions chapter. It is likely that the next section on Libraries will also join the "Modules" chapter. The question is whether "Sections" and "Modules" should be on separate pages or in a shared section on a single page.

The third proposal in this PR is to move the section on `Function` to a new section on funind in the "Libraries and plugins" chapter (since funind is a plugin that may lose its central place in the future as it is progressively replaced by Equations). The goal is to consolidate the rest of the documentation on what this plugin proposes (`functional induction` tactic in particular) into this section.

The fourth proposal in this PR is more straightforward. It is to take out the very long section on implicit arguments and to make it a section on a separate page in the new "Language extensions" chapter. Note that this section contains a sub-section on canonical structures that will need to be moved out to the "Canonical Structures" chapter at some point.

Note that I've not yet split the sections that need to be consolidated immediately with a pre-existing chapter (extended pattern matching, modules, coercions in particular).

cc @mattam82 @jfehrle 

Can you confirm that this plan sound good to you and consistent with the CEP?

### CODE REVIEW

**Comment by cpitclaudel** [comment]
A technical tip: you can split a file into multiple parts while preserving its history with git; the way to do that is to create on branch per part, then move the entire original file to the file corresponding to each new part on each branch commit, and finally remove all but the relevant text from each moved file.  Then go back to master and do an octopus merge (`git merge <all branches here>`).  This will keep blame and region histories working.

```
$ cd /tmp/
$ mkdir example
$ cd example
$ git init
Initialized empty Git repository in /tmp/example/.git/
$ printf 'x\nx\nx\nx\nx\ny\ny\ny\ny\ny\nz\nz\nz\nz\nz\n' > whole
$ git add whole
$ git commit -a -m "Init"
[master (root-commit) 261a451] Init
 1 file changed, 15 insertions(+)
 create mode 100644 whole
$ printf 'x\nx\nx\nx\nx\ny\ny\ny2\ny\ny\nz\nz\nz\nz\nz\n' > whole
$ git commit -a -m "Make a change"
[master c2eaf61] Make a change
 1 file changed, 1 insertion(+), 1 deletion(-)
$ git checkout -b part1 master
Switched to a new branch 'part1'
$ git mv whole part1
$ git commit -a -m "Prepare part 1"
[part1 373e25e] Prepare part 1
 1 file changed, 0 insertions(+), 0 deletions(-)
 rename whole => part1 (100%)
$ printf 'x\nx\nx\nx\nx\n' > part1
$ git commit -a -m "Split part 1"
[part1 c3b3cdb] Split part 1
 1 file changed, 10 deletions(-)
$ git checkout -b part2 master
Switched to a new branch 'part2'
$ git mv whole part2
$ git commit -a -m "Prepare part 2"
[part2 ba424ca] Prepare part 2
 1 file changed, 0 insertions(+), 0 deletions(-)
 rename whole => part2 (100%)
$ printf 'y\ny\ny2\ny\ny\n' > part2
$ git commit -a -m "Split part 2"
[part2 6c77ee4] Split part 2
 1 file changed, 10 deletions(-)
$ git checkout -b part3 master
Switched to a new branch 'part3'
$ git mv whole part3
$ git commit -a -m "Prepare part 3"
[part3 8db10ef] Prepare part 3
 1 file changed, 0 insertions(+), 0 deletions(-)
 rename whole => part3 (100%)
$ printf 'z\nz\nz\nz\nz\n' > part3
$ git commit -a -m "Split part 3"
[part3 627b0e4] Split part 3
 1 file changed, 10 deletions(-)
$ git checkout master
Switched to branch 'master'
$ git merge -m "Split whole into part1, part2, part3" part1 part2 part3
Fast-forwarding to: part1
Trying simple merge with part2
Trying simple merge with part3
Merge made by the 'octopus' strategy.
 part1 |  5 +++++
 part2 |  5 +++++
 part3 |  5 +++++
 whole | 15 ---------------
 4 files changed, 15 insertions(+), 15 deletions(-)
 create mode 100644 part1
 create mode 100644 part2
 create mode 100644 part3
 delete mode 100644 whole
$ git branch -d part1 part2 part3
Deleted branch part1 (was c3b3cdb).
Deleted branch part2 (was 6c77ee4).
Deleted branch part3 (was 627b0e4).
$ git log -L3,3:part2 --pretty=oneline | cat
c2eaf61ef55d8f634d354abd4a5d2f8ba91ec656 Make a change

diff --git a/whole b/whole
--- a/whole
+++ b/whole
@@ -8,1 +8,1 @@
-y
+y2
261a4519dfd151f3fe8e960f7a7a070b47923f00 Init

diff --git a/whole b/whole
--- /dev/null
+++ b/whole
@@ -0,0 +8,1 @@
+y
$ 
```

**Review by jfehrle — APPROVED** [review]
Looks good to me.  It makes sense to split up these large files.  Are these "final" based on what we know now, or somewhat preliminary to facilitate more shuffling of the material?

The PR description could use some bullet points.  It took a bit of close attention to see what's getting moved where.  It would be helpful if there was a checked-in file showing the full outline in which we could see where things are added and removed--but of course that would be created and maintained by hand and is more work.

> Can you confirm that this plan sound good to you and consistent with the CEP?

These seem like good changes--though not cast in stone.  We may come up with different and better ideas later.  Getting the logic of how the material is organized in introductions will make it clearer.

The CEP envisions evolution of the organization/outline it presented.  If we want to track our changes against the CEP outline, then maybe we should check-in the original outline and update it as we proceed.

Some suggestions and thoughts below.

File: doc/sphinx/using/libraries/funind.rst [hunk 1]
```diff
@@ -0,0 +1,169 @@
+Functional induction
+====================
+
+.. _advanced-recursive-functions:
+
+Advanced recursive functions
```
[the remark this answers is not in the document]
Comment (Zimmi48) [reply doc/sphinx/using/libraries/funind.rst]: The plan is to add more such subsections (on the `functional induction` tactic in particular).

File: doc/sphinx/language/gallina-extensions.rst [hunk 2]
```diff
@@ -6,318 +6,6 @@ Extensions of |Gallina|
 |Gallina| is the kernel language of |Coq|. We describe here extensions of
```
[the remark this answers is not in the document]
Comment (Zimmi48) [reply doc/sphinx/language/gallina-extensions.rst]: The goal is to strip this chapter apart and eventually remove it entirely. Same thing for the Galina chapter. The new "core language" and "language extensions" chapters are meant to replace them, but on several pages.

[the remark this answers is not in the document]
Comment (Zimmi48) [reply doc/sphinx/language/gallina-extensions.rst]: The goal of this PR is only to do the split, that's why I voluntarily refrained from adding any new text (beside a title).

**Comment by Zimmi48** [comment]
I'll try to add a preliminary outline using comments.

**Comment by Zimmi48** [comment]
@cpitclaudel I've updated the PR to use your octopus trick. For the parts that will need (in further PRs) to be split out only to be merged together with other parts, I suppose no similar trick exists, does it?

**Comment by cpitclaudel** [comment]
> I've updated the PR to use your octopus trick. 

Neat. This looks good to me; are we waiting for further reviews or should I merge?

> For the parts that will need (in further PRs) to be split out only to be merged together with other parts, I suppose no similar trick exists, does it?

You'd do it in two steps: first the split with the octopus trick, then two renames followed by a merge to do the concatenation.   That second step is explained here: https://devblogs.microsoft.com/oldnewthing/20190514-00/?p=102493.

**Comment by Zimmi48** [comment]
> Neat. This looks good to me; are we waiting for further reviews or should I merge?

Please proceed. I'll answer the other questions of @jfehrle in another PR (under preparation) so that the outline can be discussed more in depth.

**Review by cpitclaudel — APPROVED** [review]

**Pull request merged**